File: C++/Binary Trees/Revision_2025/BST/ValidateBST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // Every Node should be in a particular range to be a valid BST
    bool isValidBST(TreeNode* root) {
        return helper(root, LLONG_MIN, LLONG_MAX);
    }

    bool helper(TreeNode* root, auto minVal, auto maxVal){
        if (root==NULL) return true;

        // Strictly check if lies in range otherwise return statement executed
        if (root->val <= minVal || root->val >= maxVal) return false;
        
        return (helper(root->left, minVal, root->val) && helper(root->right, root->val, maxVal));
    }
};

File: C++/LKND/Validate_BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Remember that left side values are smaller than right values
# Pass Min and Max values
# If current node value <= min allowed value or >= max allowed value -> BST condition violated

class Solution:
    def isValidBST(self, root: TreeNode | None) -> bool:
        
        def helperBST(node, min_val, max_val):
            if node == None:
                return True
            
            if node.val <= min_val or node.val >= max_val:
                return False

            return (helperBST(node.left, min_val, node.val) and 
                    helperBST(node.right, node.val, max_val))

        return helperBST(root, float('-inf'), float('inf'))
